src/grpc.cc

```diff
@@ -254,168 +258,124 @@ static grpc::Status unmarshal_value_list(const collectd::types::ValueList &msg,
 } /* unmarshal_value_list() */
 
 /*
- * request call-backs and call objects
+ * Collectd service
  */
-
-static grpc::Status Process(grpc::ServerContext *ctx,
-		DispatchValuesRequest request, DispatchValuesReply *reply)
-{
-	value_list_t vl = VALUE_LIST_INIT;
-	auto status = unmarshal_value_list(request.values(), &vl);
-	if (!status.ok())
-		return status;
-
-	if (plugin_dispatch_values(&vl))
-		status = grpc::Status(grpc::StatusCode::INTERNAL,
-				grpc::string("failed to enqueue values for writing"));
-	return status;
-} /* Process(): DispatchValues */
-
-static grpc::Status Process(grpc::ServerContext *ctx,
-		QueryValuesRequest request, QueryValuesReply *reply)
-{
-	uc_iter_t *iter;
-	char *name = NULL;
-
-	value_list_t matcher;
-	auto status = unmarshal_ident(request.identifier(), &matcher, false);
-	if (!status.ok())
-		return status;
-
-	if ((iter = uc_get_iterator()) == NULL) {
-		return grpc::Status(grpc::StatusCode::INTERNAL,
-				grpc::string("failed to query values: cannot create iterator"));
-	}
-
-	status = grpc::Status::OK;
-	while (uc_iterator_next(iter, &name) == 0) {
-		value_list_t res;
-		if (parse_identifier_vl(name, &res) != 0) {
-			status = grpc::Status(grpc::StatusCode::INTERNAL,
-					grpc::string("failed to parse identifier"));
-			break;
+class CollectdImpl : public collectd::Collectd::Service {
+public:
+	grpc::Status QueryValues(grpc::ServerContext *ctx, QueryValuesRequest const *req, grpc::ServerWriter<QueryValuesResponse> *writer) override {
+		value_list_t match;
+		auto err = unmarshal_ident(req->identifier(), &match, false);
+		if (!err.ok()) {
+			return err;
 		}
 
-		if (!ident_matches(&res, &matcher))
-			continue;
-
-		if (uc_iterator_get_time(iter, &res.time) < 0) {
-			status = grpc::Status(grpc::StatusCode::INTERNAL,
-					grpc::string("failed to retrieve value timestamp"));
-			break;
+		std::queue<value_list_t> value_lists;
+		err = this->queryValuesRead(&match, &value_lists);
+		if (err.ok()) {
+			err = this->queryValuesWrite(ctx, writer, &value_lists);
```

Comment:
Yeah, also the implementation of the AVL tree doesn't allow for releasing the lock while holding an iterator, so while this would be nice, it's far out :\